Implement hof-build

lennym opened the pull request:
Performs standard build tasks for a hof project based on default configuration in a single command.

Includes a `watch` command, which will perform a granular rebuild on file changes depending on what files have changed. This will also restart server processes on changes.

Making this granular makes it *way* faster than firearms existing nodemon solution.

A default build does the following things:

* browserify compile
* sass compile
* copy images
* compile translations

## Example

Based on firearms package.json, will turn this:

```
scripts: {
    "start": "node .",
    "dev": "npm-run-all --parallel watch:app watch:scss watch:js",
    "watch:app": "NODE_ENV=development nodemon -x 'npm run hof-transpile && npm start' --ignore 'apps/**/translations/en/'",
    "watch:scss": "nodemon -e scss -x 'npm run sass'",
    "watch:js": "nodemon --watch assets/js -e js -x 'npm run browserify'",
    "test": "NODE_ENV=test npm run lint && npm run test:unit",
    "test:unit": "istanbul cover _mocha",
    "test:acceptance": "funkie so-acceptance --steps",
    "lint": "npm-run-all --parallel lint:app lint:acceptance",
    "lint:app": "eslint .",
    "lint:acceptance": "eslint --no-ignore -c ./node_modules/so-acceptance/.eslintrc ./apps/*/features/**/*.js",
    "copy:images": "cp -r ./assets/images ./public/",
    "browserify": "browserify ./assets/js/index.js > ./public/js/bundle.js",
    "create:public": "mkdir -p ./public/js ./public/css ./public/images",
    "sass": "npm-sass ./assets/scss/app.scss > ./public/css/app.css",
    "hof-transpile": "hof-transpiler ./apps/**/translations/src -w --shared ./apps/common/translations/src",
    "prepareapp": "npm run create:public; npm run sass; npm run browserify; npm run copy:images; npm run hof-transpile",
    "postinstall": "bash -c 'if [[ ${NODE_ENV} != production ]]; then npm run prepareapp; fi;'"
}
```

Into this:

```
scripts: {
    "start": "node .",
    "build": "hof-build",
    "dev": "hof-build watch",
    "test": "NODE_ENV=test npm run lint && npm run test:unit",
    "test:unit": "istanbul cover _mocha",
    "test:acceptance": "funkie so-acceptance --steps",
    "lint": "npm-run-all --parallel lint:app lint:acceptance",
    "lint:app": "eslint .",
    "lint:acceptance": "eslint --no-ignore -c ./node_modules/so-acceptance/.eslintrc ./apps/*/features/**/*.js",
    "postinstall": "bash -c 'if [[ ${NODE_ENV} != production ]]; then npm run build; fi;'"
}
```

As well as removing all associated dependencies.

lennym committed:
Implement hof-build

Performs standard build tasks for a hof project based on default configuration in a single command.

Includes a `watch` command, which will perform a granular rebuild on file changes depending on what files have changed. This will also restart server processes on changes.

easternbloc commented on bin/hof-build:
```diff
@@ -0,0 +1,5 @@
+
+const options = require('minimist')(process.argv.slice(2));
+
+require('../')(options);
```
newline?

sulthan-ahmed replied:
looks like it's missing an end of file line

easternbloc reviewed (approved):
apart from the new line i'm very much in favour of this 👍

lennym committed:
Fix linting errors in bin file

Also add eslint of that file to test command to catch future errors.

lennym committed:
Clean up index file

Moves requires out of the function, and makes default config immutable - merging onto a per-execution settings object.

lennym committed:
Allow tasks to be disabled if config set to falsy value
